src/CommonLib/LDAPUtils.cs

```diff
@@ -874,38 +873,79 @@ public IEnumerable<ISearchResultEntry> QueryLDAP(LDAPQueryOptions options)
                 }catch (LdapException le) when (le.ErrorCode == (int)LdapErrorCodes.ServerDown &&
                                                 retryCount < MaxRetries)
                 {
+                    /*A ServerDown exception indicates that our connection is no longer valid for one of many reasons.
+                    However, this function is generally called by multiple threads, so we need to be careful in recreating
+                    the connection. Using a semaphore, we can ensure that only one thread is actually recreating the connection
+                    while the other threads that hit the ServerDown exception simply wait. The initial caller will hold the semaphore
+                    and do a backoff delay before trying to make a new connection which will replace the existing connection in the 
+                    _ldapConnections cache. Other threads will retrieve the new connection from the cache instead of making a new one
+                    This minimizes overhead of new connections while still fixing our core problem.
+                    
+                    A CurrentCount of 0 indicates the semaphore is already held because c# semaphores are weird and backwards*/
+                    var isSemaphoreHeld = _semaphoreSlim.CurrentCount == 0;
```

Comment:
This state is checked before the Semaphore is acquired, so it would be subject to a race condition where another thread could acquire the semaphore first. If that happens, then we would sleep and recreate a new connection despite just having done so. Or for the other way around it would try to use the bad cached connection again.

It would be better to track who is in charge of recreating the connection inside the semaphore some way, or  maybe use a different primitive. I couldn't figure out which C# primitive could be used in the way I was looking for, but I am thinking of some construct something more like this pseudocode:
```js
// try to acquire sync primitive (atomic op). This is nonblocking
held = primitive.TryAcquire()
if (held) {
  // we are first
  conn = delayandreconnect()
  // let everyone else thru
  primitive.Release()
else {
  // someone else already has the primitive. wait for the other thread to complete
  primitive.Wait()
  // the above Wait normally blocks until released, but will also return
  // immediately if Release has been called by another thread in between
  // here and the above TryAcquire

  // at this point the other thread has done the reconnection
  conn = getexistingconnection()

  // release acquired primitive (may be unnecessary depending on type)
  primitive.Release()
}
```

Reply:
Oh, I meant to say that perhaps a `Mutex` or a `ManualResetEvent|ManualResetEventSlim` might be more like the above construct? It's hard to tell from the docs though and someone more familiar with C# might know better.

Reply:
I replaced the semaphore with a ManualResetEvent, I think you're right it suits this use case better